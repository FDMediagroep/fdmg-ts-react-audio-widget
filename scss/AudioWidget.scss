@import "~@fdmg/ts-react-progress-bar/dist/ProgressBar";