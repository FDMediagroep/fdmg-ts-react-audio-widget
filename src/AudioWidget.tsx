import * as React from 'react';
import ProgressBar from "@fdmg/ts-react-progress-bar";

export interface Props {
    playerSrc: string;
    excludeProgressBar?: boolean;
    hideProgressBarCurrentTime?: boolean;
    hideProgressBarDuration?: boolean;
    onCanPlay?: (...args: any[]) => void;
    onEnded?: (...args: any[]) => void;
    onLoadStart?: (...args: any[]) => void;
    onLoad?: (...args: any[]) => void;
    onLoadedData?: (...args: any[]) => void;
    onPause?: (...args: any[]) => void;
    onPlay?: (...args: any[]) => void;
    onPlaying?: (...args: any[]) => void;
    onSeeked?: (...args: any[]) => void;
    onSeeking?: (...args: any[]) => void;
    onSuspend?: (...args: any[]) => void;
    onTimeUpdate?: (...args: any[]) => void;
}

/**
 * AudioWidget composition - Renders a HTML5 audio element and a progress bar.
 */
export default class AudioWidget extends React.Component<Props, any> {
    public state: any;
    public props: Props;

    private audioPlayer: HTMLAudioElement;
    private progressBar: any;

    constructor(props: Props) {
        super(props);

        this.state = {
            autoPlay: false,
            buffering: false,
            currentTime: 0,
            duration: 0,
            hideTimeline: false,
            percentage: 0
        };

        this.setInitialAudioState = this.setInitialAudioState.bind(this);
        this.handleOnTimeUpdate = this.handleOnTimeUpdate.bind(this);
        this.handleUpdateElapsedTime = this.handleUpdateElapsedTime.bind(this);

        // this.setState({ hideTimeline: !this.state.autoPlay && this.isSafariMobile() ? true : false });
    }

    get audioElement() {
        return this.audioPlayer;
    }

    handleOnCanPlay = (...args) => {
        if (this.props.onCanPlay) { this.props.onCanPlay(...args); }
        this.setInitialAudioState();
    }

    /**
     * Set the initial state of the audio player
     */
    setInitialAudioState() {
        this.setState({
            currentTime : this.audioPlayer.currentTime,
            duration : this.audioPlayer.duration,
            percentage : this.calculateElapsedPercentage(this.audioPlayer.currentTime, this.audioPlayer.duration)
        });
    }

    /**
     * Calculates the elapsed percentage of audio.
     * currentTime and duration are sent by the audio element.
     */
    calculateElapsedPercentage = (currentTime: number, duration: number) => {
        const percentage = currentTime / duration * 100;
        return percentage.toFixed(2);
    }

    /**
     * Calculates the elapsed time from the given percentage.
     * @param percentage
     * @returns {number}
     */
    getElapsedTimeFromPercentage = (percentage) => {
        return (this.audioPlayer.duration / 100) * percentage;
    }

    /**
     * Callback for when the progressbar elapsed time is changed through user-interaction. Given is e.target.value
     * which is the track's percentage. It calculates the elapsed time from this percentage and updates the player controls.
     */
    handleUpdateElapsedTime(e) {
        this.audioPlayer.currentTime = this.getElapsedTimeFromPercentage(e.target.value);
        this.play();
    }

    /**
     * On every audio element time update event the current time and percentage state will be updated
     * with a new value and thus triggers a re-render of the current time counter and progress bar.
     */
    handleOnTimeUpdate(...args) {
        if (this.props.onTimeUpdate) { this.props.onTimeUpdate(...args); }
        this.setState({
            currentTime : this.audioPlayer.currentTime,
            percentage : this.calculateElapsedPercentage(this.audioPlayer.currentTime, this.audioPlayer.duration)
        });
    }

    handleOnPause = (...args) => {
        if (this.props.onPause) { this.props.onPause(...args); }
    }

    handleOnPlay = (...args) => {
        if (this.props.onPlay) { this.props.onPlay(...args); }
    }

    handleOnPlaying = (...args) => {
        if (this.props.onPlaying) { this.props.onPlaying(...args); }
    }

    handleOnSuspend = (...args) => {
        if (this.props.onSuspend) { this.props.onSuspend(...args); }
    }

    play = () => {
        this.setState({autoPlay: true});
        this.audioPlayer.play();
    }

    pause = () => {
        this.audioPlayer.pause();
    }

    buffering = () => {
        this.setState({ buffering : true });
    }

    doneBuffering = () => {
        this.setState({ buffering : false });
    }

    handleOnEnded = (...args) => {
        if (this.props.onEnded) { this.props.onEnded(...args); }
    }

    handleOnLoadStart = (...args) => {
        if (this.props.onLoadStart) { this.props.onLoadStart(...args); }
        this.buffering();
    }

    handleOnLoad = (...args) => {
        if (this.props.onLoad) { this.props.onLoad(...args); }
        this.doneBuffering();
    }

    handleOnLoadedData = (...args) => {
        if (this.props.onLoadedData) { this.props.onLoadedData(...args); }
        this.doneBuffering();
    }

    handleOnSeeking = (...args) => {
        if (this.props.onSeeking) { this.props.onSeeking(...args); }
        this.buffering();
    }

    handleOnSeeked = (...args) => {
        if (this.props.onSeeked) { this.props.onSeeked(...args); }
        this.doneBuffering();
    }

    /**
     * Returns true when on Safari mobile.
     * @returns {RegExpMatchArray}
     */
    isSafariMobile = () => {
        return (navigator.userAgent.match(/iPad/i) || navigator.userAgent.match(/iPhone/i));
    }

    render() {
        const hideTimeLine = !this.state.autoPlay && this.isSafariMobile() ? true : false;
        const progressBarComponent = this.props.excludeProgressBar ? "" : (
            <ProgressBar
                ref={(progressBar) => { this.progressBar = progressBar; }}
                currentTime={this.state.currentTime}
                duration={this.state.duration}
                percentage={this.state.percentage}
                onElapsedTimeUpdate={this.handleUpdateElapsedTime}
                autoPlay={this.state.autoPlay}
                isBuffering={this.state.buffering}
                hideTimeLine={hideTimeLine}
                hideProgressBarCurrentTime={this.props.hideProgressBarCurrentTime}
                hideProgressBarDuration={this.props.hideProgressBarDuration}
            />
        );

        return (
            <div className="audio">
                <audio
                    ref={(audioPlayer) => { this.audioPlayer = audioPlayer; }}
                    src={this.props.playerSrc}
                    onCanPlay={this.handleOnCanPlay}
                    onEnded={this.handleOnEnded}
                    onTimeUpdate={this.handleOnTimeUpdate}
                    autoPlay={this.state.autoPlay}
                    onLoadStart={this.handleOnLoadStart}
                    onLoad={this.handleOnLoad}
                    onLoadedData={this.handleOnLoadedData}
                    onPause={this.handleOnPause}
                    onPlay={this.handleOnPlay}
                    onPlaying={this.handleOnPlaying}
                    onSeeking={this.handleOnSeeking}
                    onSeeked={this.handleOnSeeked}
                    onSuspend={this.handleOnSuspend}
                />
                {progressBarComponent}
            </div>
        );
    }
}
