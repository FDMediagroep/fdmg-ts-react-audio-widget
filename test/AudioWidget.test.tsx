import * as Enzyme from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import * as React from 'react';
import AudioWidget from '../src/AudioWidget';

console.info = () => {};

describe('AudioWidget', () => {
    const empty = () => {};

    beforeAll(() => {
        Enzyme.configure({ adapter: new Adapter() });
    });

    test('renders correctly', () => {
        const component = shallow(
            <AudioWidget
                playerSrc={'test.mp3'}
                onEnded={empty}
                onTimeUpdate={empty}
                onSeeked={empty}
                onLoadedData={empty}
                onLoadStart={empty}
                onCanPlay={empty}
                onLoad={empty}
            />);
        const audioEl = component.find('audio');
        expect(toJson(component)).toMatchSnapshot();
        expect(audioEl.prop('src')).toBe('test.mp3');
    });

    test('renders correctly on Safari mobile', () => {
        const oldUserAgent = navigator.userAgent;
        Object.defineProperty(navigator, 'userAgent', {
            writable: true,
            value: 'iPad'
        });
        const component = mount(
            <AudioWidget
                playerSrc={'test.mp3'}
                onEnded={empty}
                onTimeUpdate={empty}
            />);
        expect(toJson(component)).toMatchSnapshot();
        Object.defineProperty(navigator, 'userAgent', {
            writable: true,
            value: oldUserAgent
        });
    });

    test('calculates track percentage correctly', () => {
        const audioComponent = shallow(
            <AudioWidget
                playerSrc={'test.mp3'}
                onEnded={empty}
                onTimeUpdate={empty}
            />);
        expect(audioComponent.instance().calculateElapsedPercentage(13, 133)).toBe("9.77");
    });

    test('converts seconds to human readable time', () => {
        const audioComponent = shallow(
            <AudioWidget
                playerSrc={'test.mp3'}
                onEnded={empty}
                onTimeUpdate={empty}
            />);

        expect(audioComponent.instance().convertToReadableTime(11234)).toBe('3u 7m 14s');
    });

    test('initial audio state is called by canplay event', () => {
        const spy = jest.spyOn(AudioWidget.prototype, 'setInitialAudioState');
        const audioComponent = mount(
            <AudioWidget
                playerSrc={'test.mp3'}
                onEnded={empty}
                onTimeUpdate={empty}
            />);
        const audioEl = audioComponent.find('audio');
        audioEl.simulate('canplay');
        expect(spy).toHaveBeenCalled();
    });

    test('handleOnTimeUpdate is called by timeupdate event', () => {
        const spy = jest.spyOn(AudioWidget.prototype, 'handleOnTimeUpdate');
        const audioComponent = mount(
            <AudioWidget
                playerSrc={'test.mp3'}
                onEnded={empty}
                onTimeUpdate={empty}
            />);
        const audioEl = audioComponent.find('audio');

        audioEl.simulate('timeupdate');
        expect(spy).toHaveBeenCalled();
    });

    test('displays loader on buffering/seeking', () => {
        const audioComponent = mount(
            <AudioWidget
                playerSrc={'test.mp3'}
                onEnded={empty}
                onTimeUpdate={empty}
            />);
        const audioEl = audioComponent.find('audio');

        audioEl.simulate('seeking');
        expect(audioComponent.state('buffering')).toBe(true);

        audioEl.simulate('seeked');
        expect(audioComponent.state('buffering')).toBe(false);
    });

    test('updates track time when progress bar is changed', () => {
        const spy = jest.spyOn(AudioWidget.prototype, 'handleUpdateElapsedTime');
        const audioComponent = mount(
            <AudioWidget
                playerSrc={'test.mp3'}
                onEnded={empty}
                v            />);

        audioComponent.find('audio');

        audioComponent.find('input').simulate('change', { target: { value: 75 }});
        expect(spy).toHaveBeenCalled();
    });

    test('can calculate elapsed time from percentage', () => {
        const audioComponent = mount(
            <AudioWidget
                playerSrc={'test.mp3'}
                onEnded={empty}
                onTimeUpdate={empty}
            />);

        Object.defineProperty(audioComponent.instance().audioPlayer, 'duration', {
            writable: true,
            value: 333
        });

        expect(audioComponent.instance().getElapsedTimeFromPercentage(33)).toBe(109.89);
    });

    test('callback functions are called', () => {
        const spy1 = jest.fn();
        const spy2 = jest.fn();
        const spy3 = jest.fn();
        const spy4 = jest.fn();
        const spy5 = jest.fn();
        const spy6 = jest.fn();
        const spy7 = jest.fn();
        const spy8 = jest.fn();
        const audioComponent = shallow(
            <AudioWidget
                playerSrc={'test.mp3'}
                onLoad={spy1}
                onCanPlay={spy2}
                onLoadStart={spy3}
                onLoadedData={spy4}
                onSeeked={spy5}
                onEnded={spy6}
                onSeeking={spy7}
                onTimeUpdate={spy8}
            />);
        audioComponent.instance().audioPlayer = {
            currentTime: 0,
            duration: 0
        };
        audioComponent.instance().handleOnLoad();
        expect(spy1).toHaveBeenCalled();
        audioComponent.instance().handleOnCanPlay();
        expect(spy2).toHaveBeenCalled();
        audioComponent.instance().handleOnLoadStart();
        expect(spy3).toHaveBeenCalled();
        audioComponent.instance().handleOnLoadedData();
        expect(spy4).toHaveBeenCalled();
        audioComponent.instance().handleOnSeeked();
        expect(spy5).toHaveBeenCalled();
        audioComponent.instance().handleOnEnded();
        expect(spy6).toHaveBeenCalled();
        audioComponent.instance().handleOnSeeking();
        expect(spy7).toHaveBeenCalled();
        audioComponent.instance().handleOnTimeUpdate();
        expect(spy8).toHaveBeenCalled();
        expect(spy1).toHaveBeenCalledTimes(1);
        expect(spy2).toHaveBeenCalledTimes(1);
        expect(spy3).toHaveBeenCalledTimes(1);
        expect(spy4).toHaveBeenCalledTimes(1);
        expect(spy5).toHaveBeenCalledTimes(1);
        expect(spy6).toHaveBeenCalledTimes(1);
        expect(spy7).toHaveBeenCalledTimes(1);
        expect(spy8).toHaveBeenCalledTimes(1);
    });

});
