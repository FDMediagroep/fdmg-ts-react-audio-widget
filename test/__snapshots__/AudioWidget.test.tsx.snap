// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AudioWidget renders correctly 1`] = `
<div
  className="audio"
>
  <audio
    autoPlay={false}
    onCanPlay={[Function]}
    onEnded={[Function]}
    onLoad={[Function]}
    onLoadStart={[Function]}
    onLoadedData={[Function]}
    onPause={[Function]}
    onPlay={[Function]}
    onPlaying={[Function]}
    onSeeked={[Function]}
    onSeeking={[Function]}
    onSuspend={[Function]}
    onTimeUpdate={[Function]}
    src="test.mp3"
  />
  <ProgressBar
    autoPlay={false}
    currentTime={0}
    duration={0}
    hideTimeLine={null}
    isBuffering={false}
    onElapsedTimeUpdate={[Function]}
    percentage={0}
  />
</div>
`;

exports[`AudioWidget renders correctly on Safari mobile 1`] = `
<AudioWidget
  onEnded={[Function]}
  onTimeUpdate={[Function]}
  playerSrc="test.mp3"
>
  <div
    className="audio"
  >
    <audio
      autoPlay={false}
      onCanPlay={[Function]}
      onEnded={[Function]}
      onLoad={[Function]}
      onLoadStart={[Function]}
      onLoadedData={[Function]}
      onPause={[Function]}
      onPlay={[Function]}
      onPlaying={[Function]}
      onSeeked={[Function]}
      onSeeking={[Function]}
      onSuspend={[Function]}
      onTimeUpdate={[Function]}
      src="test.mp3"
    />
    <ProgressBar
      autoPlay={false}
      currentTime={0}
      duration={0}
      hideTimeLine={
        Array [
          "iPad",
        ]
      }
      isBuffering={false}
      onElapsedTimeUpdate={[Function]}
      percentage={0}
    >
      <div
        className="progress"
      >
        <div />
        <div
          className="counters"
        >
          <span />
          <span />
        </div>
      </div>
    </ProgressBar>
  </div>
</AudioWidget>
`;
